beacon-chain/rpc/beaconv1/blocks.go

```diff
@@ -24,20 +118,234 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
+		}
+		if root == nil {
+			return nil, status.Errorf(codes.Internal, "No head root was found")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: root,
+			},
+		}, nil
+	case "finalized":
+		finalized := bs.FinalizationFetcher.FinalizedCheckpt()
+		if finalized == nil {
+			return nil, status.Errorf(codes.Internal, "No finalized root was found")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: finalized.Root,
+			},
+		}, nil
+	case "genesis":
+		blk, err := bs.BeaconDB.GenesisBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for genesis slot: %v", err)
+		}
+		if blk == nil {
+			return nil, status.Error(codes.Internal, "Could not find genesis block")
+		}
+		root, err := blk.Block.HashTreeRoot()
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not hash genesis block")
+		}
+		return &ethpb.BlockRootResponse{
+			Data: &ethpb.BlockRootContainer{
+				Root: root[:],
+			},
+		}, nil
+	default:
+		if len(req.BlockId) == 32 {
+			return &ethpb.BlockRootResponse{
+				Data: &ethpb.BlockRootContainer{
+					Root: req.BlockId,
+				},
+			}, nil
+		} else {
+			slot := bytesutil.BytesToUint64BigEndian(req.BlockId)
+			roots, err := bs.BeaconDB.BlockRoots(ctx, filters.NewFilter().SetStartSlot(slot).SetEndSlot(slot))
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for slot %d: %v", slot, err)
+			}
+
+			numRoots := len(roots)
+			if numRoots == 0 {
+				return nil, nil
+			}
+			root := roots[0]
+			return &ethpb.BlockRootResponse{
+				Data: &ethpb.BlockRootContainer{
+					Root: root[:],
+				},
+			}, nil
+		}
+	}
 }
 
 // ListBlockAttestations retrieves attestation included in requested block.
 func (bs *Server) ListBlockAttestations(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockAttestationsResponse, error) {
-	return nil, errors.New("unimplemented")
+	blk, err := bs.blockFromBlockID(ctx, req.BlockId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block from block ID: %v", err)
+	}
+	if blk == nil {
+		return nil, status.Errorf(codes.Internal, "Could not find requested block")
+	}
+
+	v1Block, err := v1alpha1ToV1Block(blk)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert block to v1")
+	}
+	return &ethpb.BlockAttestationsResponse{
+		Data: v1Block.Block.Body.Attestations,
+	}, nil
+}
+
+func (bs *Server) blockFromBlockID(ctx context.Context, blockId []byte) (*ethpb_alpha.SignedBeaconBlock, error) {
+	var err error
+	var blk *ethpb_alpha.SignedBeaconBlock
+	switch string(blockId) {
+	case "head":
+		blk, err = bs.HeadFetcher.HeadBlock(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not retrieve head block")
+		}
+		if blk == nil {
+			return nil, errors.New("no head block was found")
```

Comment:
Shouldn't you return nil, nil here and let the caller understand that blockFromBlockID didn't find anything? That way you can more appropriately return codes.NotFound / 404.